### PR TITLE
Rosetta SDK: Initial Draft Implementation

### PR DESCRIPTION
**Pull request by SmartArray**
# Introduction
This PR contains an initial draft implementation for Coinbase's official Rosetta Spec, written in NodeJS/Javascript.

## State of this implementation
The state of this implementation is still _an early alpha_. All components have been ported from the official Go reference implementation to maintain 100% compatibility.

**At the moment: the following unit tests have been ported to this implementation:**

- syncer/syncer_test.go ✅
- fetcher/account_test.go ✅
- fetcher/network_test.go ✅
- fetcher/block_test.go ✅
- asserter/block_test.go ✅
- asserter/account_test.go ✅
- asserter/network_test.go ✅
- asserter/request_test.go ✅
- asserter/asserter_test.go ✅
- asserter/construction_test.go ✅
- asserter/request.go ✅
- asserter/construction.go ✅
- reconciler/reconciler_test.go ✅
- parser/balance_changes_test.go ✅
- parser/group_operations_test.go ✅
- parser/match_operations_test.go ✅
 
## Additional Remarks
1) We have used the official OpenAPI NodeJS Express Generator to generate the server stub.
2) Furthermore, the Client Module was generated with the OpenAPI Javascript Generator. This generator was executed twice to have one promised version, and one with callbacks. They have also generated type classes, which are located in ```lib/client/src/model```. These may be exported to another separate module in the future.

3) Documentation has been added for Fetcher Module only. See ```lib/fetcher/README.md```.

### CODE REVIEW

**Review by gto90 — APPROVED** [review]

**Comment by ChillingSilence** [comment]
Tested successfully with:
  17 passing (1s)
And second lot:
  136 passing (104ms)

File: lib/client/src/api/promises/NetworkApi.js [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+      };
+      let headerParams = {
+      };
+      let formParams = {
+      };
+
+      let authNames = [];
+      let contentTypes = ['application/json'];
+      let accepts = ['application/json'];
+      let returnType = NetworkStatusResponse;
```
Comment (19h) [inline lib/client/src/api/promises/NetworkApi.js]: Is this code generated? Those orphaned param objects look somewhat off :-)

Comment (gto90) [reply lib/client/src/api/promises/NetworkApi.js]: Yes, this code is autogenerated by the OpenAPI 3.0 generator.

**Comment by gto90** [comment]
@SmartArray , can we close this PR out in favor of the newer one?

File: examples/server/index.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+const RosettaSDK = require('../..');
+
+// Create a server configuration
+const Server = new RosettaSDK.Server({
+  URL_PORT: 8080,
+});
+
+Server.launch()
+  .catch(e => logger.error(e));
```
Comment (gto90) [inline examples/server/index.js]: @SmartArray ,

Can you add a new line here?  I'll take a look at the project structure but we should be using ES Lint with Semistandard rules.  Most IDEs would catch this as a result and offer to automatically fix them.

File: index.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  Server: RosettaServer,
+  Reconciler: RosettaReconciler,
+  Syncer: RosettaSyncer,
+  Fetcher: RosettaFetcher,
+  Client: RosettaClient,
+  Parser: RosettaParser,
+
+  RosettaSyncerEvents,
+  Errors,
+};
```
Comment (gto90) [inline index.js:57]: Same here @SmartArray , We'll want to use something like ES Lint to fix all this up automatically.

**Review by gto90 — CHANGES_REQUESTED** [review]
@SmartArray ,

This is looking really good.  Just a few comments on implementing ESLint with Semistandard rules and fixing up some formatting issues here and there.

File: examples/server/index.js (hunk 2 above)
Comment (SmartArray) [reply examples/server/index.js]: Correct.

File: index.js (hunk 3 above)
Comment (SmartArray) [reply index.js:57]: Thought I was using ESLint, but let me double check.

**Comment by SmartArray** [comment]
> @SmartArray , can we close this PR out in favor of the newer one?

Kk, let us close all the PRs. Will do one for every version. Please create a branch 1.3.1 and 1.4.1 so I can PR against them.

**Pull request merged**